query/src/interpreters/async_insert_queue.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::hash::Hash;
+use std::sync::Arc;
+
+use common_base::base::tokio::sync::Notify;
+use common_base::base::tokio::time::interval;
+use common_base::base::tokio::time::Duration;
+use common_base::base::tokio::time::Instant;
+use common_base::base::ProgressValues;
+use common_base::base::Runtime;
+use common_base::infallible::Mutex;
+use common_base::infallible::RwLock;
+use common_datablocks::DataBlock;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_planners::InsertPlan;
+use common_planners::SelectPlan;
+
+use super::InsertInterpreter;
+use super::SelectInterpreter;
+use crate::pipelines::new::executor::PipelineCompleteExecutor;
+use crate::pipelines::new::processors::port::InputPort;
+use crate::pipelines::new::processors::port::OutputPort;
+use crate::pipelines::new::processors::BlocksSource;
+use crate::pipelines::new::SinkPipeBuilder;
+use crate::pipelines::new::SourcePipeBuilder;
+use crate::sessions::QueryContext;
+use crate::sessions::SessionManager;
+use crate::sessions::SessionType;
+use crate::sessions::Settings;
+use crate::storages::memory::MemoryTableSink;
+
+#[derive(Clone)]
+pub struct InsertKey {
+    plan: Arc<InsertPlan>,
+    // settings different with default settings
+    changed_settings: Settings,
+}
+
+impl InsertKey {
+    pub fn get_serialized_changed_settings(&self) -> String {
+        let mut serialized_settings = String::new();
+        let values = self.changed_settings.get_setting_values();
+        for value in values.into_iter() {
+            let serialized = serde_json::to_string(&value).unwrap();
```

Comment:
Why do we need to compare settings? Serializing `InsertKey` every time will bring extra CPU overhead.